macros/graph/plotly3D.pl

```diff
@@ -0,0 +1,626 @@
+      [0.7, 'rgb(230,145,90)'], [1, 'rgb(178,10,28)']]"
+
+A colorset can have any number of color points between 0 and 1. To make
+the plot a single color, set the color for 0 and 1 to be the same:
+
+    "[[0, 'rgb(0,200,0)'], [1, 'rgb(0,200,0)']]"
+
+=head1 FUNCTION TYPES
+
+The functions to generate the plot can be either mathematical, JavaScript,
+Perl, or raw data, and this can be controlled using the funcType => type
+option in addCurve or addSurface methods. The valid types are:
+
+  jsmd    This is the default type, in which the functions are converted
+          from math formulas into JavaScript functions to generate the
+          plot. This should accept standard mathematical notation with
+          some exceptions: Multiplication must be an explicit "*".
+          "ucos(v)" is not be accepted, but "u*cos(v)" will. JavaScript
+          considers "-u^2" not well defined, instead use "-(u^2)".
+
+  js      The functions are interpreted as raw JavaScript functions. The
+          functions will be passed the defined variables and return a
+          single value.
+
+  perl    The functions are interpreted as Perl subroutines. The functions
+          will be passed the appropriate number of inputs, and return a
+          single value.
+
+  data    The functions are interpreted as a nested array of data points to
+          be sent directly to plotly to plot. This is useful for static plots
+          in which the points do not need to be generated each time.
+
+=head1 Graph3D OPTIONS
+
+Create a graph object: $graph = Graph3D(option => value)
+The valid options are:
+
+  height      The height and width of the div containing the graph.
+  width
+
+  title       Graph title to print above the graph.
+
+  style       CSS style to style the div containing the graph.
+
+  bgcolor     The background color of the graph.
+
+  image       Image filename to be used in hardcopy TeX output.
+              If no image is provided, the hardcopy TeX output
+              has a message that image must be viewed online.
+
+  tex_size    Size of image in hardcopy TeX output as scale factor from 0 to 1000.
+              1000 is 100%, 500 is 50%, etc.
+
+  tex_border  Put (1) or don't put (0) a border around image in TeX output.
+
+  scene       Add a JavaScript scene configuration dictionary to the plotly layout.
+              Example: scene => 'aspectmode: "manual", aspectratio: {x: 1, y: 1, z: 1},
+              xaxis: { range: [0,2] }, yaxis: { range: [0,3] }, zaxis: { range: [1,4] }'
+              https://plotly.com/javascript/3d-axes/ for more examples.
+
+=cut
+
+sub _plotly3D_init {
+	ADD_JS_FILE('https://cdn.plot.ly/plotly-latest.min.js', 1);
+	PG_restricted_eval("sub Graph3D {new plotly3D(\@_)}");
+}
+
+our $plotlyCount = 0;
+
+package plotly3D;
+
+sub new {
+	my $self  = shift;
+	my $class = ref($self) || $self;
+
+	$plotlyCount++;
+	$self = bless {
+		id         => $plotlyCount,
```

Comment:
Using a simple count for html ids is certainly not good enough.  If multiple problems use this macro on the same page, then this will create id clashes.

Reply:
You should use `$main::PG->getUniqueName('plotly3D')` to get an id that will be unique for this problem, and different  from any other problems that might be on the page that also use the macro.  That uses `PG_alias` and incorporates things into the returned name that make sure this is the case.

Reply:
This didn't quite work, as the name includes `-` which won't work in variable names, but `$main::PG->getUniqueName('plotly3D') =~ s/-/_/gr` seems to work under my initial tests.

Reply:
Yes, you need to do the `s/-/_/gr`.  I forgot to mention that.